src/activity.rs

```diff
@@ -11,107 +11,151 @@ use crate::frame::*;
   width: usize,
-  height: usize,
-  // Side of unit (square) activity block in log2
   granularity: usize,
+
+  pub tot_bins: usize,
+  pub seg_bins: [usize; 8],
+  pub avg_var: f64,
 }
 
 impl ActivityMask {
   #[hawktracer(activity_mask_from_plane)]
-  pub fn from_plane<T: Pixel>(luma_plane: &Plane<T>) -> ActivityMask {
-    let PlaneConfig { width, height, .. } = luma_plane.cfg;
+  pub fn from_plane<T: Pixel>(fi: &FrameInvariants<T>, luma_plane: &Plane<T>) -> ActivityMask {
+    let granularity: usize = 3; /* Granularity of the map */
+    let act_granularity: usize = 4; /* Granularity of the activity data, if != granularity it repeats */
+    let tot_pix: usize = (1 << granularity) * (1 << granularity);
+
+    /* Aligned width and height to the activity data */
+    let width = (((luma_plane.cfg.width >> act_granularity) << act_granularity)) + (1 << act_granularity);
+    let height = (((luma_plane.cfg.height >> act_granularity) << act_granularity)) + (1 << act_granularity);
 
-    let granularity = 3;
+    let mut variances =
+      Vec::with_capacity((width >> granularity) * (height >> granularity));
+    variances.resize((width >> granularity) * (height >> granularity), 0f64);
 
     let aligned_luma = Rect {
       x: 0_isize,
       y: 0_isize,
-      width: (width >> granularity) << granularity,
-      height: (height >> granularity) << granularity,
+      width: width,
+      height: height,
     };
     let luma = PlaneRegion::new(luma_plane, aligned_luma);
 
-    let mut variances =
-      Vec::with_capacity((height >> granularity) * (width >> granularity));
+    let mut freq_storage: Aligned<[T::Coeff; 64 * 64]> = Aligned::uninitialized();
+    let mut src_storage: Aligned<[i16; 64 * 64]> = Aligned::uninitialized();
+
+    let mut avg_var = 0f64;
+    let mut max = 0f64;
 
-    for y in 0..height >> granularity {
-      for x in 0..width >> granularity {
+    for y in 0..height >> act_granularity {
+      for x in 0..width >> act_granularity {
         let block_rect = Area::Rect {
-          x: (x << granularity) as isize,
-          y: (y << granularity) as isize,
-          width: 8,
-          height: 8,
+          x: (x << act_granularity) as isize,
+          y: (y << act_granularity) as isize,
+          width: 1 << act_granularity,
+          height: 1 << act_granularity,
         };
 
         let block = luma.subregion(block_rect);
 
-        let mean: f64 = block
-          .rows_iter()
-          .flatten()
-          .map(|&pix| {
-            let pix: i16 = CastFromPrimitive::cast_from(pix);
-            pix as f64
-          })
-          .sum::<f64>()
-          / 64.0_f64;
-        let variance: f64 = block
-          .rows_iter()
-          .flatten()
-          .map(|&pix| {
-            let pix: i16 = CastFromPrimitive::cast_from(pix);
-            (pix as f64 - mean).powi(2)
-          })
-          .sum::<f64>();
-        variances.push(variance);
+        let cpu = CpuFeatureLevel::default();
```

Comment:
You can use `fi.cpu_feature_level` to avoid the overhead of looking up the supported CPU features.